plugins/net_plugin/net_plugin.cpp

```diff
@@ -4040,92 +3948,265 @@ namespace eosio {
+      std::lock_guard g( connections_mtx );
+      if( find_connection_impl( host ) )
+         return "already connected";
+
+      connect_impl( host );
       return "added connection";
    }
 
-   string net_plugin_impl::disconnect( const string& host ) {
-      std::lock_guard<std::shared_mutex> g( connections_mtx );
-      for( auto itr = connections.begin(); itr != connections.end(); ++itr ) {
-         if( (*itr)->peer_address() == host ) {
-            fc_ilog( logger, "disconnecting: ${cid}", ("cid", (*itr)->connection_id) );
-            (*itr)->close();
-            connections.erase(itr);
-            return "connection removed";
-         }
+   string connections_manager::disconnect( const string& host ) {
+      std::lock_guard g( connections_mtx );
+      if( auto c = find_connection_impl( host ) ) {
+         fc_ilog( logger, "disconnecting: ${cid}", ("cid", c->connection_id) );
+         c->close();
+         connections.erase(c);
+         supplied_peers.erase(host);
+         return "connection removed";
       }
       return "no known connection for host";
    }
 
-   string net_plugin::disconnect( const string& host ) {
-      return my->disconnect(host);
+   void connections_manager::close_all() {
+      fc_ilog( logger, "close all ${s} connections", ("s", connections.size()) );
+      std::lock_guard g( connections_mtx );
+      for( auto& con : connections ) {
+         fc_dlog( logger, "close: ${cid}", ("cid", con->connection_id) );
+         con->close( false, true );
+      }
+      connections.clear();
    }
 
-   std::optional<connection_status> net_plugin::status( const string& host )const {
-      std::shared_lock<std::shared_mutex> g( my->connections_mtx );
-      auto con = my->find_connection( host );
-      if( con )
+   connection_ptr connections_manager::round_robin_next( const connection_ptr& current, uint32_t sync_known_lib_num ) const {
+      connection_ptr new_sync_source = current;
+      std::shared_lock g( connections_mtx );
+      if( connections.empty() ) {
+         new_sync_source.reset();
+      } else if( connections.size() == 1 ) {
+         if (!new_sync_source) {
+            new_sync_source = *connections.begin();
+         }
+      } else {
+         // init to a linear array search
+         auto cptr = connections.begin();
+         auto cend = connections.end();
+         // do we remember the previous source?
+         if (current) {
+            //try to find it in the list
+            cptr = connections.find( current );
+            cend = cptr;
+            if( cptr == connections.end() ) {
+               // not there - must have been closed! cend is now connections.end, so just flatten the ring.
+               new_sync_source.reset();
+               cptr = connections.begin();
+            } else {
+               // was found - advance the start to the next. cend is the old source.
+               if( ++cptr == connections.end() && cend != connections.end() ) {
```

Comment:
At this point we know that `cend != connections.end()`:

```suggestion
               if( ++cptr == connections.end() ) {
```

Reply:
Removed